File: apps/server/src/domain/diagram/workspace_diagrams.rs
use async_trait::async_trait;

use super::{Diagram, DiagramDescription, DraftEdge, DraftNode};
use crate::domain::{HasMany, ServerError};

#[async_trait]
pub trait WorkspaceDiagrams: HasMany<Diagram> {
    async fn add(&self, desc: DiagramDescription) -> Result<Diagram, ServerError>;

    async fn update(
        &self,
        diagram_id: &str,
        desc: DiagramDescription,
    ) -> Result<Diagram, ServerError>;

    async fn delete(&self, diagram_id: &str) -> Result<(), ServerError>;

    async fn list(&self, page: u32, page_size: u32) -> Result<(Vec<Diagram>, u64), ServerError>;

    async fn save_diagram(
        &self,
        diagram_id: &str,
        draft_nodes: Vec<DraftNode>,
        draft_edges: Vec<DraftEdge>,
    ) -> Result<(), ServerError>;

    async fn publish_diagram(&self, diagram_id: &str) -> Result<(), ServerError>;
}

File: libs/server-nest/domain/src/domain/diagram/workspace-diagrams.ts
import { HasMany } from '../core';
import { Diagram, DiagramDescription } from './diagram';
import { DraftEdge, DraftNode } from './types';

export interface WorkspaceDiagrams extends HasMany<Diagram> {
  add(desc: DiagramDescription): Promise<Diagram>;
  update(diagramId: string, desc: DiagramDescription): Promise<Diagram>;
  delete(diagramId: string): Promise<void>;
  list(page: number, pageSize: number): Promise<[Diagram[], number]>;
  saveDiagram(
    diagramId: string,
    draftNodes: DraftNode[],
    draftEdges: DraftEdge[],
  ): Promise<void>;
}
